server/src/main/java/org/elasticsearch/index/snapshots/blobstore/SlicedInputStream.java

```diff
@@ -99,4 +110,38 @@ public final int available() throws IOException {
         return stream == null ? 0 : stream.available();
     }
 
+    @Override
+    public boolean markSupported() {
+        return true;
+    }
+
+    @Override
+    public void mark(int readLimit) {
+        // We ignore readLimit since openSlice() can re-open previous InputStreams, and we can skip as many bytes as we'd like.
+        // According to JDK documentation, marking a closed InputStream should have no effect.
+        if (closed == false) {
+            if (initialized && nextSlice > 0) { // nextSlice > 0 guards the case it is initialized and numSlices is 0.
+                markedSlice = nextSlice - 1;
+                markedSliceOffset = currentSliceOffset;
+            } else {
+                markedSlice = 0;
+                markedSliceOffset = 0;
+            }
+        }
+    }
+
+    @Override
+    public void reset() throws IOException {
+        // JDK documentation does not clarify if a closed InputStream (that has been previously marked) can be reset. We assume the same
+        // behavior specified for mark(), i.e., that reset on a closed InputStream has no effect.
+        if (closed == false) {
+            if (markedSlice < 0 || markedSliceOffset < 0) {
+                throw new IOException("Mark has not been set");
+            }
+
+            nextSlice = markedSlice;
+            nextStream();
+            skipNBytes(markedSliceOffset);
```

Comment:
I find it safer to call `skipNBytes` on the inputstream returned from `nextStream` so that we can be sure the skipping does not trigger `nextStream` again due to any bug.

Reply:
Good idea! Done.